### PR TITLE
fix(explorer): make preview work for unsaved explorers

### LINKED ISSUE #1818: Explorer preview doesn't work with `grapherId`-type views
## Description

In the explorer editor, the live preview will not show a preview of a `grapherId` until the explorer is saved (which means it will be deployed).

## Expected behaviour

The live preview should show grapherIds just as well as table-based explorers.

## Steps to reproduce

1. Go to this URL: http://localhost:3030/admin/explorers/new
2. Paste the below spreadsheet
3. Look at the live preview
4. Save and then look at the live preview again

```tsv
explorerTitle	My New Explorer
graphers
	grapherId	Statistic Radio
	64	Life Expectancy
	225	GDP per capita
```

## Technical context

This is caused by the architecture of the live preview.
The live preview is purely based on the explorer config as it is present in `localStorage`.
However, for `grapherId`-based views there is an additional server-side step to provide the required grapher configs inline: https://github.com/owid/owid-grapher/blob/a39b489aef03453c0e2fbab27de72e32510df651/baker/siteRenderers.tsx#L527-L531

This step, for obvious reasons, cannot happen for explorer configs that are only known at the client side.

## Additional context

@pabloarosado reported this [on Slack](https://owid.slack.com/archives/C46U9LXRR/p1670925162434509)

### PR DESCRIPTION
**Pull request by sophiamersmann**
fixes #1818

### Problem

- The live preview does not work for unsaved explorers that refer to grapher config IDs or rely on partial grapher configs
   - applies to new explorers that have never been saved
   - applies to existing explorers with unsaved changes

- Core problem:
    - Grapher configs are baked into the explorer page in a server-side step – this only happens when the explorer is saved
    - If not saved, these configs are missing in the live preview, meaning that nothing or only a partial explorer can be shown

- This has been known to be a problem for a while; however, note that the bug has a different quality for indicator-based explorers and partial grapher configs:
    - in the case of grapher id-based explorers, the preview of a new explorer does not show anything until saved
    - in the case of indicator-based explorers, the preview of an unsaved explorer does show _something_ but that _something_ might be incorrect as partial grapher configs have not been applied yet

### Proposed solution 

- The live preview parses available inlined configs and fetches missing configs if necessary

Not optimal since:
- Live preview and actual explorers sometimes have different code paths
    - Although to publish an explorer, it must be saved so that an author will see the "actual" preview eventually (by "actual" preview, I mean a preview that does not rely on additional requests to get missing configs)
- Fetching from the admin endpoints in `Explorer.tsx` does not feel clean 

### Alternatives

- Do not bake grapher configs and always fetch them late -> degrades user experience since rendering the chart title immediately wouldn't be possible
- Store one "new" explorer per user and bake configs of grapher IDs before saving -> seems weird and complicated
- Continue baking configs the way we do but "patch up" the live preview by fetching configs that are missing -> this is what I opted for

---

I don't know how often authors work on unsaved explorers. Also, although indicator-based explorers can theoretically suffer from this problem, they won't for some time because we don't have any partial grapher configs at the moment. We could leave this PR as a draft and revisit it when there is a more pressing need to solve this problem. Happy to discuss this!

### CODE REVIEW

**Comment by sophiamersmann** [comment]
Current dependencies on/for this PR:
* master
  * **PR #2332** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/2332" target="_blank">Graphite</a> 
    * **PR #2333** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/2333" target="_blank">Graphite</a> 
      * **PR #2349** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/2349" target="_blank">Graphite</a> 
        * **PR #2394** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/2394" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/2394?utm_source=stack-comment).

**Comment by sophiamersmann** [comment]
Current dependencies on/for this PR:
* master
  * **PR #2332** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/2332" target="_blank">Graphite</a> 
    * **PR #2333** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/2333" target="_blank">Graphite</a> 
      * **PR #2349** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/2349" target="_blank">Graphite</a> 
        * **PR #2394** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/2394" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/2394?utm_source=stack-comment).

File: explorer/Explorer.tsx [hunk 1]
```diff
@@ -87,18 +88,85 @@ export interface ExplorerProps extends SerializedGridProgram {
     selection?: SelectionArray
 }
 
+const grapherConfigPath = (chartId: number): string =>
+    `${ADMIN_BASE_URL}/admin/api/charts/${chartId}.config.json`
+const partialGrapherConfigPath = (variableId: number): string =>
+    `${ADMIN_BASE_URL}/admin/api/variables/${variableId}.grapherConfig.json`
+
+async function fetchJSON<T>(url: string): Promise<T> {
```
Comment (sophiamersmann) [inline explorer/Explorer.tsx:96]: We should probably be using the [`getJSON` function in `Admin.tsx`](https://github.com/owid/owid-grapher/blob/330ee415038ed4d233d36e2c7d7d99022065ddb6/adminSiteClient/Admin.tsx#L178), but I didn't want there to be an explicit link between the Admin and the Explorer

File: adminSiteServer/apiRouter.ts [hunk 2]
```diff
@@ -1770,6 +1770,27 @@ apiRouter.get(
...
+            `SELECT id, grapherConfig FROM variables WHERE id=?`,
+            [variableId]
+        )
+
+        if (!variable) {
+            throw new JsonError(`No variable by id '${variableId}'`, 404)
+        }
+
+        const config = JSON.parse(variable.grapherConfig)
+        config.id = variable.id
```
[the remark this answers is not in the document]
Comment (sophiamersmann) [reply adminSiteServer/apiRouter.ts:1789]: I am also using the id field in indicator-based explorers, https://github.com/owid/owid-grapher/blob/2d66314ac3cd480bbc6af3c89a8a6198bb245ecc/baker/siteRenderers.tsx#L631 i.e. for grapher configs that are curated on the indicator level, the config id is the indicator id. That felt natural to me but I can see how that could potentially lead to problems. It's not too late to change it!

(p.s. if charts are created on the fly, then their id is set to zero): https://github.com/owid/owid-grapher/blob/2d66314ac3cd480bbc6af3c89a8a6198bb245ecc/explorer/Explorer.tsx#L388

**Comment by danyx23** [comment]
@sophiamersmann was it intentional to let this be closed without merging?

**Comment by sophiamersmann** [comment]
Yes and no. I'd love to see this issue fixed (esp. now that authors start using indicator-based explorers with partial grapher configs) but it's not obvious what the best solution is and since I didn't have time in the past weeks, I decided to let it go for now.

**Pull request closed without merging**